tilavarauspalvelu/api/graphql/types/unit/filtersets.py

```diff
@@ -44,6 +46,28 @@ def filter_by_only_with_permission(self, qs: models.QuerySet, name: str, value:
 
         return qs.filter(Q(id__in=u_ids) | Q(unit_groups__in=g_ids)).distinct()
 
+    # These filters use information across the relationship between Unit and ReservationUnit.
+    # FilterSets apply individual filters in separate `queryset.filter(...)` calls.
+    #
+    # Due to how Django works when spanning `to-many` relationships in qs.filter(...) calls,
+    # (see https://docs.djangoproject.com/en/5.0/topics/db/queries/#spanning-multi-valued-relationships),
+    # this means that using a combination of these filters would result in a queryset where
+    # ANY of their conditions are true for ANY ReservationUnit linked to a given Unit.
+    #     e.g. using `published_reservation_units=True` and `only_direct_bookable=True`
+    #     would return all Units where ANY of their ReservationUnit are EITHER published OR directly bookable.
+    #
+    # In this case, this is incorrect, since we want less permissive behavior:
+    #     e.g. the aforementioned filter should return all Units where ANY of their ReservationUnit are
+    #     BOTH published AND directly bookable.
+    #
+    # The incorrect behavior also has the side effect of adding multiple SQL joins to the
+    # many-to-many though table between Unit and ReservationUnit, which makes the query
+    # slower, since more duplication is needed.
+    #
+    # To prevent this, we need these filters to execute in order, and set `queryset.query.filter_is_sticky = True`
+    # to indicate to the queryset that it should reuse joins it found from one filter to the next.
+    # Note that this only applies until the next time the queryset is cloned, (e.g. when using
+    # `queryset.filter(...)` or `.distinct()`) and thus needs to be reapplied between them.
```

Comment:
This explanation text is now in the wrong place since it instructs that the _filters_, not the functions, need to be defined in sequence (to be executed in sequence), meaning these:

```
published_reservation_units = django_filters.BooleanFilter(method="filter_by_published_reservation_units")
own_reservations = django_filters.BooleanFilter(method="filter_by_own_reservations")
only_direct_bookable = django_filters.BooleanFilter(method="filter_by_only_direct_bookable")
only_seasonal_bookable = django_filters.BooleanFilter(method="filter_by_only_seasonal_bookable")
```

Obviously, you don't want to include it twice for both subclasses, but it should be referenced there somehow.

Reply:
Added a comment to both subclasses to refer here.